design-proposals/storage-migration/storage-migration.md

```diff
@@ -0,0 +1,282 @@
+        // the destination PVC. The claim name will contain the prefix plus a
+        // random string.
+        PrefixDestinationVolume string `json:"PrefixDestinationVolume" valid:"required"`
+        // ReclaimPolicySourcePvc describes how the source volumes will be
+        // treated after a successful migration
+        ReclaimPolicySourcePvc ReclaimPolicySourcePvc `json:"reclaimPolicySourcePvc,omitempty"`
+}
+
+// ReclaimPolicySourcePvc describes how the source PVC will be treated after the storage migration completes. The policies follows the same behavior as the RetainPolicy for PVs: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reclaiming
+type ReclaimPolicySourcePvc string
+
+const (
+        DeleteReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Delete"
+        RetainReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Retain"
+)
+
+// StorageMigrationSpec is the spec for a StorageMigration resource
+type StorageMigrationSpec struct {
+        // The name of the VMI to perform the migration on. VMI must exist in the storage migration objects
+        // namespace
+        VMIName string `json:"vmiName,omitempty" valid:"required"`
+        // MigrationVolumes is a list of volumes to migrate
+        MigrationVolumes []MigrationVolume `json:"migrationVolumes,omitempty"`
+        // MigrationStorageClass is a list containing the information for migrating the volumes of
+        // a source storage class to another destination storage class
+        MigrationStorageClasses []MigrationStorageClass `json:"migrationStorageClasses,omitempty"`
+
+}
+
+// StorageMigrationStatus is the status for a StorageMigration resource
+type StorageMigrationStatus struct {
+        MigrationVolumes []MigrationVolume `json:"migrationVolumes,omitempty"`
+}
+```
+
+There are two methods for specifying the volumes that will be transferred.
+
+The manual list of the source and target PVCs is the first available option. In this scenario, the user must have already created the destination PVCs.
+
+The second choice is to merely define the storage classes for the source and destination. Based on the selected class and source PVC specification, KubeVirt will create the appropriate destination PVCs for this option. The user needs to specify the prefix name to be used for the destination PVCs created by KubeVirt, and a random string will be append to every PVCs.
+
+The user could mix `MigrationVolumes` and MigrationStorageClasses`. For example a volumes of the source storage class specified in MigrationStorageClasses could be present also in `MigrationVolumes`. In this case, the MigrationVolumes takes the precedence and the destination volume will be created by KubeVirt with the information specified for the DestinationPVC. This can be a practical way how the user could overwrite the defaults for a storage class just for a single volume. E.g for overwriting the default naming or the ReclaimPolicySourcePvc.
```

Comment:
Confused by this:  `the destination volume will be created by KubeVirt with the information specified for the DestinationPVC`

With `MigrationVolumes`, the user is expected to create the target PVC, correct?  I would expect KubeVirt TO NOT create the target PVC if the source volume was encountered via `MigrationStorageClasses`.

If `MigrationStorageClasses` has label selectors I think it may be fine to have it be mutually exclusive with `MigrationVolumes`

Reply:
This was suggested by @vasiliy-ul in https://github.com/kubevirt/community/pull/245#discussion_r1360347469. I might have badly explained what we agreed.

There is a potential conflict between using `MigrationVolumes` and `MigrationStorageClasses`. A PVC could be listed under the `MigrationVolumes` but the same time its storage class be included in the `MigrationStorageClasses`.

Initially, I was thinking to reject such cases, but instead, Vasily suggested something smarter. 

We could use this case for tuning certain specifications of a single volume. For example, if I want to overwrite the naming for a specific volume, then I can simply list the PVC under the `MigrationVolumes` and the controller instead of using the naming convention specified for the storage class, it will use the name of the volume under `MigrationVolumes`.

So, yes the user is expected to manually create the PVC IF there aren't any MigrationStorageClasses specified or any of the storage classes specified there don't correspond to the source PVC class.

I'll try to make this clearer in the proposal, but please let me know if you like the idea!